ops/_main.py

```diff
@@ -0,0 +1,555 @@
+            )
+            # Note that we don't exit nonzero, because that would cause Juju to rerun the hook
+            raise _Abort(0)
+
+        if self._use_juju_for_storage:
+            store = ops.storage.JujuStorage()
+        else:
+            store = ops.storage.SQLiteStorage(charm_state_path)
+        return store
+
+    def _make_framework(self, dispatcher: _Dispatcher):
+        # If we are in a RelationBroken event, we want to know which relation is
+        # broken within the model, not only in the event's `.relation` attribute.
+
+        if self._juju_context.dispatch_path.endswith('-relation-broken'):
+            broken_relation_id = self._juju_context.relation_id
+        else:
+            broken_relation_id = None
+
+        model = ops.model.Model(
+            self._charm_meta, self._model_backend, broken_relation_id=broken_relation_id
+        )
+        store = self._make_storage(dispatcher)
+        framework = ops.framework.Framework(
+            store,
+            self._charm_root,
+            self._charm_meta,
+            model,
+            event_name=dispatcher.event_name,
+            juju_debug_at=self._juju_context.debug_at,
+        )
+        framework.set_breakpointhook()
+        return framework
+
+    def _emit(self):
+        """Emit the event on the charm."""
+        # TODO: Remove the collect_metrics check below as soon as the relevant
+        #       Juju changes are made. Also adjust the docstring on
+        #       EventBase.defer().
+        #
+        # Skip reemission of deferred events for collect-metrics events because
+        # they do not have the full access to all hook tools.
+        if not self.dispatcher.is_restricted_context():
+            # Re-emit any deferred events from the previous run.
+            self.framework.reemit()
+
+        # Emit the Juju event.
+        _emit_charm_event(self.charm, self.dispatcher.event_name, self._juju_context)
+        # Emit collect-status events.
+        ops.charm._evaluate_status(self.charm)
+
+    def _commit(self):
+        """Commit the framework and gracefully teardown."""
+        self.framework.commit()
+
+    def run(self):
+        """Emit and then commit the framework."""
+        try:
+            self._emit()
+            self._commit()
+        finally:
+            self.framework.close()
+
+
+def main(charm_class: Type[ops.charm.CharmBase], use_juju_for_storage: Optional[bool] = None):
+    """Set up the charm and dispatch the observed event.
+
+    The event name is based on the way this executable was called (argv[0]).
+
+    .. jujuremoved:: 4.0
+        The ``use_juju_for_storage`` argument is not available from Juju 4.0
+
+    Args:
+        charm_class: the charm class to instantiate and receive the event.
+        use_juju_for_storage: whether to use controller-side storage. If not specified
+            then Kubernetes charms that haven't previously used local storage and that
```

Comment:
This is wrong in the current ops docs, but while we're here, could we fix this? The "Kubernetes charms" is no longer correct. It's only the case for podspec charms, not Kubernetes sidecar charms, which is the default K8s choice now.

Reply:
done